pkg/storage/unified/sql/notifier.go

```diff
@@ -0,0 +1,118 @@
+package sql
+
+import (
+	"context"
+	"sync"
+
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/storage/unified/resource"
+	"github.com/grafana/grafana/pkg/storage/unified/sql/db"
+	"github.com/grafana/grafana/pkg/storage/unified/sql/dbutil"
+	"github.com/grafana/grafana/pkg/storage/unified/sql/sqltemplate"
+)
+
+type eventNotifier interface {
+	notify(ctx context.Context) (<-chan *resource.WrittenEvent, error)
+	// send will forward an event to all subscribers who want to be notified.
+	//
+	// Note: depending on the implementation, send might be noop and new events
+	//       will be fetched from an external source.
+	send(ctx context.Context, event *resource.WrittenEvent)
+	close()
+}
+
+func newNotifier(b *backend) (eventNotifier, error) {
+	if b.isHA {
+		b.log.Info("Using polling notifier")
+		notifier, err := newPollingNotifier(&pollingNotifierConfig{
+			pollingInterval: b.pollingInterval,
+			watchBufferSize: b.watchBufferSize,
+			log:             b.log,
+			tracer:          b.tracer,
+			batchLock:       b.batchLock,
+			listLatestRVs:   b.listLatestRVs,
+			historyPoll: func(ctx context.Context, grp string, res string, since int64) ([]*historyPollResponse, error) {
+				var records []*historyPollResponse
+				err := b.db.WithTx(ctx, ReadCommittedRO, func(ctx context.Context, tx db.Tx) error {
+					var err error
+					records, err = dbutil.Query(ctx, tx, sqlResourceHistoryPoll, &sqlResourceHistoryPollRequest{
+						SQLTemplate:          sqltemplate.New(b.dialect),
+						Resource:             res,
+						Group:                grp,
+						SinceResourceVersion: since,
+						Response:             &historyPollResponse{},
+					})
+					return err
+				})
+				return records, err
+			},
+			done:    b.done,
+			dialect: b.dialect,
+		})
+		if err != nil {
+			return nil, err
+		}
+		return notifier, nil
+	}
+
+	b.log.Info("Using channel notifier")
+	return newChannelNotifier(b.watchBufferSize, b.log), nil
+}
+
+type channelNotifier struct {
+	subscribers map[chan *resource.WrittenEvent]bool
+	mu          sync.RWMutex
```

Comment:
Mutex is typically on the top of the group of fields it protects. (mutex hat)

Reply:
TIL

https://github.com/go-critic/go-critic/issues/465

Reply:
I'm putting the fields the hat on! TIL 👍

Reply:
Reordered the fields, so all fields under the mutex are the protected ones: d6c495a29f0ae8ded0b759b59c9c4a6c7af2ecef